chips/stm32f303xc/src/flash.rs

```diff
@@ -0,0 +1,608 @@
+    pub fn load_option(&self) {
+        self.registers.cr.modify(Control::OBLLAUNCH::SET);
+    }
+
+    pub fn handle_interrupt(&self) {
+        if self.registers.sr.is_set(Status::EOP) {
+            // Reset by writing a 1.
+            self.registers.sr.modify(Status::EOP::SET);
+
+            match self.state.get() {
+                FlashState::Write => {
+                    self.write_counter.set(self.write_counter.get() + 2);
+
+                    if self.write_counter.get() == PAGE_SIZE {
+                        self.registers.cr.modify(Control::PG::CLEAR);
+                        self.state.set(FlashState::Ready);
+                        self.write_counter.set(0);
+
+                        self.client.map(|client| {
+                            self.buffer.take().map(|buffer| {
+                                client.write_complete(buffer, hil::flash::Error::CommandComplete);
+                            });
+                        });
+                    } else {
+                        self.program_halfword();
+                    }
+                }
+                FlashState::Erase => {
+                    if self.registers.cr.is_set(Control::PER) {
+                        self.registers.cr.modify(Control::PER::CLEAR);
+                    }
+
+                    if self.registers.cr.is_set(Control::MER) {
+                        self.registers.cr.modify(Control::MER::CLEAR);
+                    }
+
+                    self.state.set(FlashState::Ready);
+                    self.client.map(|client| {
+                        client.erase_complete(hil::flash::Error::CommandComplete);
+                    });
+                }
+                FlashState::WriteOption => {
+                    self.registers.cr.modify(Control::OPTPG::CLEAR);
+                    self.state.set(FlashState::Ready);
+                }
+                FlashState::EraseOption => {
+                    self.registers.cr.modify(Control::OPTER::CLEAR);
+                    self.state.set(FlashState::Ready);
+                }
+                _ => {}
+            }
+        }
+
+        if self.state.get() == FlashState::Read {
+            self.state.set(FlashState::Ready);
+            self.client.map(|client| {
+                self.buffer.take().map(|buffer| {
+                    client.read_complete(buffer, hil::flash::Error::CommandComplete);
+                });
+            });
+        }
+
+        if self.registers.sr.is_set(Status::WRPRTERR) {
+            if self.registers.cr.is_set(Control::PG) {
+                self.registers.cr.modify(Control::PG::CLEAR);
+            }
+
+            if self.registers.cr.is_set(Control::OPTPG) {
+                self.registers.cr.modify(Control::OPTPG::CLEAR);
+            }
+
+            // Reset by writing a 1.
+            self.registers.sr.modify(Status::WRPRTERR::SET);
+            panic!("WRPRTERR: programming a write-protected address.");
```

Comment:
Writing a 1 to an already set register causes a reset? What does it reset? Also, why did you choose to panic here instead of returning an error? I assume that if the address was write protected no write actually occurred.

Reply:
My bad. What i actually meant was that if you want to clear WRPRTERR, PGERR or EOP (all the bits that are set by hardware interrupts) you actually have to write a 1 instead of a 0, hence the CLEAR instead of SET. I will change the comment to "Cleared by writing a 1".

Reply:
Got it, thanks!